### PR TITLE
Kubernetes clone server group operation

### PR DESCRIPTION
**Pull request by skim1420**
@spinnaker/google

### CODE REVIEW

File: clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/deploy/ops/CloneKubernetesAtomicOperation.groovy [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+
+    // Build description object from ancestor, override any values that were specified on the clone call
+    newDescription.application = description.application ?: ancestorNames.app
+    newDescription.stack = description.stack ?: ancestorNames.stack
+    newDescription.freeFormDetails = description.freeFormDetails ?: ancestorNames.detail
+    newDescription.targetSize = description.targetSize ?: ancestorServerGroup.spec?.replicas
+    newDescription.loadBalancers = []
+    newDescription.securityGroups = []
+    ancestorServerGroup.spec?.template?.metadata?.labels?.each { key, val ->
+      if (key.startsWith("load-balancer-")) {
```
Comment (lwander) [inline clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/deploy/ops/CloneKubernetesAtomicOperation.groovy]: You should move this logic into `KubernetesUtil`

File: clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/deploy/validators/StandardKubernetesAttributeValidator.groovy [hunk 2]
```diff
@@ -26,6 +26,7 @@ class StandardKubernetesAttributeValidator {
   static final prefixPattern = /^[a-z0-9]+$/
   static final cpuPattern = /^(\d+m)?$/
   static final memoryPattern = /^(\d+(Mi|Gi))?$/
+  static final serverGroupNamePattern = /^([a-zA-Z0-9]*-[a-zA-Z0-9]*){2,3}$/
```
Comment (lwander) [inline clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/deploy/validators/StandardKubernetesAttributeValidator.groovy]: This would validate `---`, which is not a valid name. You should probably check that a `v\d{3}` exists at the end of the string too.

Comment (duftler) [inline clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/deploy/validators/StandardKubernetesAttributeValidator.groovy]: A handful of tests for this regex would help.

File: clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/deploy/description/CloneKubernetesAtomicOperationDescription.groovy [hunk 3]
```diff
@@ -0,0 +1,28 @@
+/*
+ * Copyright 2015 Google, Inc.
```
Comment (duftler) [inline clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/deploy/description/CloneKubernetesAtomicOperationDescription.groovy]: These should all read `2016`.

File: clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/deploy/ops/CloneKubernetesAtomicOperation.groovy [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+  Void operate(List priorOutputs) {
+    CloneKubernetesAtomicOperationDescription newDescription = cloneAndOverrideDescription()
+
+    task.updateStatus BASE_PHASE, "Initializing copy of server group for " +
+      "${description.application}-${description.stack}..."
+
+    new DeployKubernetesAtomicOperation(newDescription).operate(priorOutputs)
+
+    task.updateStatus BASE_PHASE, "Finished copying server group for " +
+      "${description.application}-${description.stack}. "
```
Comment (duftler) [inline clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/deploy/ops/CloneKubernetesAtomicOperation.groovy]: Trailing space just before terminating quote.

File: clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/deploy/ops/DeployKubernetesAtomicOperation.groovy [hunk 5]
```diff
@@ -43,16 +45,16 @@ class DeployKubernetesAtomicOperation implements AtomicOperation<Void> {
...
     def credentials = description.kubernetesCredentials
     def clusterName = KubernetesUtil.combineAppStackDetail(description.application, description.stack, description.freeFormDetails)
-    task.updateStatus KubernetesUtil.BASE_PHASE, "Looking up next sequence index..."
+    task.updateStatus BASE_PHASE, "Looking up next sequence index..."
     def sequenceIndex = KubernetesUtil.getNextSequence(clusterName, credentials)
-    task.updateStatus KubernetesUtil.BASE_PHASE, "Sequence index chosen to be ${sequenceIndex}."
+    task.updateStatus BASE_PHASE, "Sequence index chosen to be ${sequenceIndex}."
     def replicationControllerName = String.format("%s-v%s", clusterName, sequenceIndex)
-    task.updateStatus KubernetesUtil.BASE_PHASE, "Replication controller name chosen to be ${replicationControllerName}"
+    task.updateStatus BASE_PHASE, "Replication controller name chosen to be ${replicationControllerName}"
```
Comment (duftler) [inline clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/deploy/ops/DeployKubernetesAtomicOperation.groovy]: Missing trailing period.

File: clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/deploy/ops/CloneKubernetesAtomicOperation.groovy [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+    TaskRepository.threadLocalTask.get()
+  }
+
+  CloneKubernetesAtomicOperationDescription description
+
+/*
+ * curl -X POST -H "Content-Type: application/json" -d  '[ { "cloneServerGroup": { "source": { "serverGroupName": "k8s-test-v000" }, "credentials":  "my-k8s-account" } } ]' localhost:7002/kubernetes/ops
+ * curl -X POST -H "Content-Type: application/json" -d  '[ { "cloneServerGroup": { "stack": "prod", "freeFormDetails": "mdservice", "targetSize": "4", "source": { "serverGroupName": "k8s-test-v000" }, "credentials":  "my-k8s-account" } } ]' localhost:7002/kubernetes/ops
+*/
+  // TODO: should return DeploymentResult object
```
Comment (duftler) [inline clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/deploy/ops/CloneKubernetesAtomicOperation.groovy]: Any reason this todo can't be addressed now?

File: clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/deploy/validators/CloneKubernetesAtomicOperationValidator.groovy [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+import com.netflix.spinnaker.clouddriver.kubernetes.deploy.description.CloneKubernetesAtomicOperationDescription
+import com.netflix.spinnaker.clouddriver.kubernetes.KubernetesOperation
+import com.netflix.spinnaker.clouddriver.orchestration.AtomicOperations
+import com.netflix.spinnaker.clouddriver.security.AccountCredentialsProvider
+import org.springframework.beans.factory.annotation.Autowired
+import org.springframework.stereotype.Component
+import org.springframework.validation.Errors
+
+@KubernetesOperation(AtomicOperations.CLONE_SERVER_GROUP)
+@Component("cloneKubernetesAtomicOperationValidator")
```
Comment (duftler) [inline clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/deploy/validators/CloneKubernetesAtomicOperationValidator.groovy]: Can drop the `("cloneKubernetesAtomicOperationValidator")`.

File: clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/KubernetesConfiguration.groovy [hunk 8]
```diff
@@ -50,4 +51,9 @@ class KubernetesConfiguration {
   KubernetesHealthIndicator kubernetesHealthIndicator() {
     new KubernetesHealthIndicator()
   }
+
+  @Bean
+  KubernetesUtil kubernetesUtil() {
```
Comment (lwander) [inline clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/KubernetesConfiguration.groovy]: I don't know spring very well, so why is this needed?

File: clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/deploy/KubernetesUtil.groovy [hunk 9]
```diff
@@ -54,6 +53,10 @@ class KubernetesUtil {
     credentials.client.replicationControllers().inNamespace(credentials.namespace).list()
   }
 
+  public ReplicationController getReplicationController(KubernetesCredentials credentials, String serverGroupName) {
```
Comment (duftler) [inline clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/deploy/KubernetesUtil.groovy]: Can drop the `public` here.

File: clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/KubernetesConfiguration.groovy (hunk 8 above)
Comment (duftler) [inline clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/KubernetesConfiguration.groovy]: Now that `KubernetesUtil` gets injected, we will want to update the `deploy` operation to have it autowired as well. Please drop a todo to make those changes and address the other `static` methods. Then we can do that work in a followon PR.

File: clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/deploy/description/CloneKubernetesAtomicOperationDescription.groovy (hunk 3 above)
Comment (skim1420) [inline clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/deploy/description/CloneKubernetesAtomicOperationDescription.groovy]: ack

File: clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/deploy/ops/CloneKubernetesAtomicOperation.groovy (hunk 4 above)
Comment (skim1420) [inline clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/deploy/ops/CloneKubernetesAtomicOperation.groovy]: ack

File: clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/deploy/ops/DeployKubernetesAtomicOperation.groovy (hunk 5 above)
Comment (skim1420) [inline clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/deploy/ops/DeployKubernetesAtomicOperation.groovy]: ack

File: clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/deploy/validators/CloneKubernetesAtomicOperationValidator.groovy (hunk 7 above)
Comment (skim1420) [inline clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/deploy/validators/CloneKubernetesAtomicOperationValidator.groovy]: ack

File: clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/KubernetesConfiguration.groovy (hunk 8 above)
Comment (skim1420) [inline clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/KubernetesConfiguration.groovy]: It allows autowiring of KubernetesUtil in, makes testing easier.

File: clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/deploy/ops/CloneKubernetesAtomicOperation.groovy [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+    task.updateStatus BASE_PHASE, "Initializing copy of server group for " +
+      "${description.application}-${description.stack}..."
+
+    new DeployKubernetesAtomicOperation(newDescription).operate(priorOutputs)
+
+    task.updateStatus BASE_PHASE, "Finished copying server group for " +
+      "${description.application}-${description.stack}. "
+  }
+
+  // Made public for testability, not intended to be called directly
```
Comment (ttomsu) [inline clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/deploy/ops/CloneKubernetesAtomicOperation.groovy]: Comment isn't really necessary - everything in groovy is more-or-less public and `private` modifier are more for documentation purposes. See http://www.gmarwaha.com/blog/2009/02/02/groovy-private-is-not-private/

File: clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/deploy/ops/CloneKubernetesAtomicOperation.groovy [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+  CloneKubernetesAtomicOperationDescription cloneAndOverrideDescription() {
+    CloneKubernetesAtomicOperationDescription newDescription = description.clone()
+
+    task.updateStatus BASE_PHASE, "Reading ancestor server group $description.source.serverGroupName..."
+
+    ReplicationController ancestorServerGroup = kubernetesUtil.getReplicationController(description.kubernetesCredentials, description.source.serverGroupName)
+
+    if (!ancestorServerGroup) {
+      throw IllegalArgumentException("Source server group $description.source.serverGroupName does not exist")
```
Comment (ttomsu) [inline clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/deploy/ops/CloneKubernetesAtomicOperation.groovy]: I'd say `IllegalStateException` is more apt

File: clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/deploy/ops/CloneKubernetesAtomicOperation.groovy [hunk 12]
```diff
@@ -0,0 +1,109 @@
...
+    // Build description object from ancestor, override any values that were specified on the clone call
+    newDescription.application = description.application ?: ancestorNames.app
+    newDescription.stack = description.stack ?: ancestorNames.stack
+    newDescription.freeFormDetails = description.freeFormDetails ?: ancestorNames.detail
+    newDescription.targetSize = description.targetSize ?: ancestorServerGroup.spec?.replicas
+    newDescription.loadBalancers = []
+    newDescription.securityGroups = []
+    ancestorServerGroup.spec?.template?.metadata?.labels?.each { key, val ->
+      if (key.startsWith("load-balancer-")) {
+        newDescription.loadBalancers.push(key.substring(14, key.length()))
```
Comment (ttomsu) [inline clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/deploy/ops/CloneKubernetesAtomicOperation.groovy]: 14 here and 15 below are magic numbers. 

1. Make them constants
1. When doing so, make it clear where it comes from, i.e.:
  ```
  public static final String LOAD_BALANCER_PREFIX_LENGTH = "load-balancer-".length()
  ```

File: clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/deploy/ops/CloneKubernetesAtomicOperation.groovy (hunk 1 above)
Comment (skim1420) [inline clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/deploy/ops/CloneKubernetesAtomicOperation.groovy]: ack

File: clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/deploy/ops/CloneKubernetesAtomicOperation.groovy (hunk 12 above)
Comment (skim1420) [inline clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/deploy/ops/CloneKubernetesAtomicOperation.groovy]: ack

File: clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/deploy/ops/CloneKubernetesAtomicOperation.groovy (hunk 11 above)
Comment (skim1420) [inline clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/deploy/ops/CloneKubernetesAtomicOperation.groovy]: ack

File: clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/deploy/ops/CloneKubernetesAtomicOperation.groovy (hunk 10 above)
Comment (skim1420) [inline clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/deploy/ops/CloneKubernetesAtomicOperation.groovy]: ack

File: clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/deploy/KubernetesUtil.groovy (hunk 9 above)
Comment (skim1420) [inline clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/deploy/KubernetesUtil.groovy]: ack

File: clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/deploy/ops/CloneKubernetesAtomicOperation.groovy (hunk 11 above)
Comment (lwander) [inline clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/deploy/ops/CloneKubernetesAtomicOperation.groovy]: Just wondering, why is `IllegalState` better than `IllegalArgument`? It seems like this exception is being thrown in response to a non-existent replication controller provided as an argument to clone.

Comment (lwander) [inline clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/deploy/ops/CloneKubernetesAtomicOperation.groovy]: @ttomsu

File: clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/deploy/validators/StandardKubernetesAttributeValidator.groovy (hunk 2 above)
Comment (skim1420) [inline clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/deploy/validators/StandardKubernetesAttributeValidator.groovy]: done

File: clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/deploy/validators/StandardKubernetesAttributeValidator.groovy [hunk 13]
```diff
@@ -26,6 +26,7 @@ class StandardKubernetesAttributeValidator {
   static final prefixPattern = /^[a-z0-9]+$/
   static final cpuPattern = /^(\d+m)?$/
   static final memoryPattern = /^(\d+(Mi|Gi))?$/
+  static final serverGroupNamePattern = /^[a-zA-Z0-9]+(-[a-zA-Z0-9]+){1,3}$/
```
Comment (lwander) [inline clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/deploy/validators/StandardKubernetesAttributeValidator.groovy]: You can have dashes in the `details` part of the `application-stack-details`.

File: clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/KubernetesConfiguration.groovy (hunk 8 above)
Comment (skim1420) [inline clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/KubernetesConfiguration.groovy]: done

File: clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/deploy/validators/StandardKubernetesAttributeValidator.groovy (hunk 13 above)
Comment (skim1420) [inline clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/deploy/validators/StandardKubernetesAttributeValidator.groovy]: done

File: clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/deploy/ops/CloneKubernetesAtomicOperation.groovy (hunk 6 above)
Comment (skim1420) [inline clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/deploy/ops/CloneKubernetesAtomicOperation.groovy]: Requires deploy to return DeploymentResult object. Will address in future PR

File: clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/deploy/ops/CloneKubernetesAtomicOperation.groovy (hunk 11 above)
Comment (skim1420) [inline clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/deploy/ops/CloneKubernetesAtomicOperation.groovy]: Throwing newly created KubernetesResourceNotFoundException

**Comment by skim1420** [comment]
@spinnaker/google addressed all comments, ptal

File: clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/KubernetesConfiguration.groovy (hunk 8 above)
Comment (skim1420) [inline clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/KubernetesConfiguration.groovy]: That is, added TODO

File: clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/deploy/ops/CloneKubernetesAtomicOperation.groovy [hunk 14]
```diff
@@ -0,0 +1,103 @@
...
+    this.description = description
+  }
+
+  private static Task getTask() {
+    TaskRepository.threadLocalTask.get()
+  }
+
+  CloneKubernetesAtomicOperationDescription description
+
+/*
```
Comment (duftler) [inline clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/deploy/ops/CloneKubernetesAtomicOperation.groovy]: Indentation is off for comment.

File: clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/deploy/validators/StandardKubernetesAttributeValidator.groovy [hunk 15]
```diff
@@ -26,6 +26,7 @@ class StandardKubernetesAttributeValidator {
   static final prefixPattern = /^[a-z0-9]+$/
   static final cpuPattern = /^(\d+m)?$/
   static final memoryPattern = /^(\d+(Mi|Gi))?$/
+  static final serverGroupNamePattern = /^[a-zA-Z0-9]+(-[a-zA-Z0-9]+)+$/
```
Comment (duftler) [inline clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/deploy/validators/StandardKubernetesAttributeValidator.groovy]: As discussed offline, I think we can drop the base server group name validation. If it's not found, it's not found. Don't see why we have to hold an opinion about the format of the base server group's name.

File: clouddriver-kubernetes/src/test/groovy/com/netflix/spinnaker/clouddriver/kubernetes/deploy/validators/CloneKubernetesAtomicOperationValidatorSpec.groovy [hunk 16]
```diff
@@ -0,0 +1,371 @@
...
+      def errorsMock = Mock(Errors)
+    when:
+      validator.validate([], description, errorsMock)
+    then:
+      1 * errorsMock.rejectValue("${DESCRIPTION}.serverGroupName", "${DESCRIPTION}.serverGroupName.invalid (Must match ${StandardKubernetesAttributeValidator.serverGroupNamePattern})")
+      0 * errorsMock._
+  }
+
+  void "validation reject (invalid source server group name 2)"() {
+    setup:
```
Comment (duftler) [inline clouddriver-kubernetes/src/test/groovy/com/netflix/spinnaker/clouddriver/kubernetes/deploy/validators/CloneKubernetesAtomicOperationValidatorSpec.groovy]: Indentation is off for `setup`/`when`/`then` blocks.

File: clouddriver-kubernetes/src/test/groovy/com/netflix/spinnaker/clouddriver/kubernetes/deploy/validators/CloneKubernetesAtomicOperationValidatorSpec.groovy [hunk 17]
```diff
@@ -0,0 +1,371 @@
...
+    def errorsMock = Mock(Errors)
+    when:
+    validator.validate([], description, errorsMock)
+    then:
+    1 * errorsMock.rejectValue("${DESCRIPTION}.serverGroupName", "${DESCRIPTION}.serverGroupName.invalid (Must match ${StandardKubernetesAttributeValidator.serverGroupNamePattern})")
+    0 * errorsMock._
+  }
+
+  void "validation reject (invalid source server group name 3)"() {
+    setup:
```
Comment (duftler) [inline clouddriver-kubernetes/src/test/groovy/com/netflix/spinnaker/clouddriver/kubernetes/deploy/validators/CloneKubernetesAtomicOperationValidatorSpec.groovy]: Indentation is off for `setup`/`when`/`then` blocks.

File: clouddriver-kubernetes/src/test/groovy/com/netflix/spinnaker/clouddriver/kubernetes/deploy/ops/CloneKubernetesAtomicOperationSpec.groovy [hunk 18]
```diff
@@ -0,0 +1,167 @@
...
+
+  }
+
+  void "builds a description based on ancestor server group, overrides nothing"() {
+    setup:
+      def inputDescription = new CloneKubernetesAtomicOperationDescription(
+        source: [serverGroupName: ANCESTOR_SERVER_GROUP_NAME]
+      )
+
+    @Subject def operation = new CloneKubernetesAtomicOperation(inputDescription)
```
Comment (duftler) [inline clouddriver-kubernetes/src/test/groovy/com/netflix/spinnaker/clouddriver/kubernetes/deploy/ops/CloneKubernetesAtomicOperationSpec.groovy]: Indentation is off here.

File: clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/deploy/ops/CloneKubernetesAtomicOperation.groovy [hunk 19]
```diff
@@ -0,0 +1,103 @@
...
+
+  CloneKubernetesAtomicOperationDescription cloneAndOverrideDescription() {
+    CloneKubernetesAtomicOperationDescription newDescription = description.clone()
+
+    task.updateStatus BASE_PHASE, "Reading ancestor server group ${description.source.serverGroupName}..."
+
+    ReplicationController ancestorServerGroup = kubernetesUtil.getReplicationController(description.kubernetesCredentials, description.source.serverGroupName)
+
+    if (!ancestorServerGroup) {
+      throw new KubernetesResourceNotFoundException("Source server group $description.source.serverGroupName does not exist")
```
Comment (duftler) [inline clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/deploy/ops/CloneKubernetesAtomicOperation.groovy]: Missing terminating period.

Comment (skim1420) [inline clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/deploy/ops/CloneKubernetesAtomicOperation.groovy]: done

File: clouddriver-kubernetes/src/test/groovy/com/netflix/spinnaker/clouddriver/kubernetes/deploy/ops/CloneKubernetesAtomicOperationSpec.groovy (hunk 18 above)
Comment (skim1420) [inline clouddriver-kubernetes/src/test/groovy/com/netflix/spinnaker/clouddriver/kubernetes/deploy/ops/CloneKubernetesAtomicOperationSpec.groovy]: done

File: clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/deploy/validators/StandardKubernetesAttributeValidator.groovy (hunk 15 above)
Comment (skim1420) [inline clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/deploy/validators/StandardKubernetesAttributeValidator.groovy]: done

File: clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/deploy/ops/CloneKubernetesAtomicOperation.groovy (hunk 14 above)
Comment (skim1420) [inline clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/deploy/ops/CloneKubernetesAtomicOperation.groovy]: done

File: clouddriver-kubernetes/src/test/groovy/com/netflix/spinnaker/clouddriver/kubernetes/deploy/validators/CloneKubernetesAtomicOperationValidatorSpec.groovy (hunk 17 above)
Comment (skim1420) [inline clouddriver-kubernetes/src/test/groovy/com/netflix/spinnaker/clouddriver/kubernetes/deploy/validators/CloneKubernetesAtomicOperationValidatorSpec.groovy]: obsolete

File: clouddriver-kubernetes/src/test/groovy/com/netflix/spinnaker/clouddriver/kubernetes/deploy/validators/CloneKubernetesAtomicOperationValidatorSpec.groovy (hunk 16 above)
Comment (skim1420) [inline clouddriver-kubernetes/src/test/groovy/com/netflix/spinnaker/clouddriver/kubernetes/deploy/validators/CloneKubernetesAtomicOperationValidatorSpec.groovy]: obsolete

**Comment by duftler** [comment]
LGTM. Please rebase and squash all commits.

**Comment by lwander** [comment]
:sparkles:
